### ISSUE TITLE
TopNode not defined

### ISSUE THREAD

**Issue by AStupidBear** [denizyuret/Knet.jl#22]
julia> using Knet
WARNING: Base.SparseMatrix is deprecated.
  likely near /home/rluser/.julia/v0.5/CUSPARSE/src/CUSPARSE.jl:2
WARNING: Method definition (::Type{Knet._CudaArray})(CUDArt.CudaArray{#T<:Any, #N<:Any}) in module Knet at /home/rluser/.julia/v0.5/Knet/src/util/cudart.jl:127 overwritten at /home/rluser/.julia/v0.5/Knet/src/util/cudart.jl:128.
WARNING: Base.writemime is deprecated.
  likely near /home/rluser/.julia/v0.5/Knet/src/util/cudart.jl:133
WARNING: Base.writemime is deprecated.
  likely near /home/rluser/.julia/v0.5/Knet/src/util/cudart.jl:133
WARNING: Base.LambdaStaticData is deprecated, use LambdaInfo instead.
  likely near /home/rluser/.julia/v0.5/Knet/src/util/deepcopy.jl:21
ERROR: LoadError: LoadError: UndefVarError: TopNode not defined
 in include_from_node1(::String) at ./loading.jl:426 (repeats 2 times)
 in eval(::Module, ::Any) at ./boot.jl:234
 in require(::Symbol) at ./loading.jl:357
while loading /home/rluser/.julia/v0.5/Knet/src/util/deepcopy.jl, in expression starting on line 21
while loading /home/rluser/.julia/v0.5/Knet/src/Knet.jl, in expression starting on line 17

**Comment by denizyuret** [comment]
Which version of Julia are you using?

**Comment by AStupidBear** [comment]
0.5rc3Yao-------- Original Message --------Subject: Re: [denizyuret/Knet.jl] TopNode not defined (#22)From: denizyuret To: "denizyuret/Knet.jl" CC: Yao Lu ,Author Which version of Julia are you using?

**Comment by denizyuret** [comment]
The current release of Knet is not yet compatible with 0.5.  I will release
a compatible version when 0.5 is officially released.

**Comment by ppalmes** [comment]
Hi,

We are testing the Julia port in PowerPC machines. Unfortunately, Julia compiles only in 0.5 Julia. Viral is asking us to test the port and I'm interested to test the GPU support of Julia using deep learning package like Knet. I have similar error above during Knet's building the CUDArt. It will be cool if Knet runs in PowerPC:

using Knet                                                             
[29/110]
INFO: Precompiling module CUDArt.
WARNING: `@windows` is deprecated, use `@static is_windows()` instead
 in depwarn(::String, ::Symbol) at ./deprecated.jl:64
 in @windows(::Any, ::Any) at ./deprecated.jl:472
 in include_from_node1(::String) at ./loading.jl:481 (repeats 2 times)
 in macro expansion; at ./none:2 [inlined]
 in anonymous at ./<missing>:?
 in eval(::Module, ::Any) at ./boot.jl:234
 in process_options(::Base.JLOptions) at ./client.jl:239
 in _start() at ./client.jl:318
while loading /localhome/paulito/.julia/v0.6/CUDArt/src/libcudart-6.5.jl, in expression starting on line 23
WARNING: Base.WORD_SIZE is deprecated.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/libcudart-6.5.jl:36
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.UTF8String is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.UTF8String is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.UTF8String is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.UTF8String is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.UTF8String is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
in _precompile_ at /localhome/paulito/.julia/v0.6/CUDArt/src/precompile.jl
WARNING: Base.UTF8String is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
in _precompile_ at /localhome/paulito/.julia/v0.6/CUDArt/src/precompile.jl
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
in _precompile_ at /localhome/paulito/.julia/v0.6/CUDArt/src/precompile.jl
WARNING: Base.UTF8String is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
in _precompile_ at /localhome/paulito/.julia/v0.6/CUDArt/src/precompile.jl
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
in _precompile_ at /localhome/paulito/.julia/v0.6/CUDArt/src/precompile.jl
WARNING: Base.ASCIIString is deprecated, use String instead.
  likely near /localhome/paulito/.julia/v0.6/CUDArt/src/CUDArt.jl:57
in _precompile_ at /localhome/paulito/.julia/v0.6/CUDArt/src/precompile.jl
WARNING: Base.SparseMatrix is deprecated.
  likely near /localhome/paulito/.julia/v0.6/CUSPARSE/src/CUSPARSE.jl:2
ERROR: LoadError: LoadError: UndefVarError: TopNode not defined
 in include_from_node1(::String) at ./loading.jl:481 (repeats 2 times)
 in eval(::Module, ::Any) at ./boot.jl:234
 in require(::Symbol) at ./loading.jl:409
while loading /localhome/paulito/.julia/v0.6/Knet/src/util/deepcopy.jl, in expression starting on line 21
while loading /localhome/paulito/.julia/v0.6/Knet/src/Knet.jl, in expression starting on line 17

**Comment by denizyuret** [comment]
I will take a look...

**Comment by denizyuret** [comment]
I switched the latest master to 0.8 development.  You should be able to
test it on 0.5 Julia:

julia> Pkg.checkout("Knet")
julia> using Knet
julia> include(Pkg.dir("Knet/examples/mnist.jl"))
julia> MNIST.main()  # trains on cpu, gpu is not activated at initialization
julia> gpu(true)
julia> MNIST.main()  # trains on gpu
julia> MNIST.main("--help")  # prints out the options

Let me know if this solves your problem.

Knet8 is under development, the documentation and updates are not up to
date, but the code should work.

best,
deniz

**Comment by ppalmes** [comment]
Thanks.

I tried running in 0.5 but CUDArt has issues. Viral sent us 0.6.0 for testing.

I did Pkg.build("Knet") after checkout and it compiled successfully in 0.6 Julia.

gpu(true) also worked but running MNIST.main() result to errors: it may indicate I'll convert Float32 to Float64?

opts=(:seed,-1)(:batchsize,100)(:hidden,Int64[])(:epochs,10)(:lr,0.5)(:atype,"KnetArray{Float32}")(:winit,0.1)(:fast,false)
ERROR: ccall: could not find function mul_32_01 in library 
 in .*(::Float32, ::Knet.KnetArray{Float32,2}) at /localhome/paulito/.julia/v0.6/Knet/src/cuda01.jl:62
 in ./(::Knet.KnetArray{Float32,2}, ::Float64) at /localhome/paulito/.julia/v0.6/Knet/src/cuda01.jl:32
 in (::MNIST.#xbatch#6{DataType,Int64,Float64})(::Array{UInt8,1}) at /localhome/paulito/.julia/v0.6/Knet/examples/mnist.jl:110
 in #minibatch#5(::Type{T}, ::Type{T}, ::Int64, ::Int64, ::Float64, ::Function, ::Array{UInt8,1}, ::Array{UInt8,1}, ::Int64) at /localhome/paulito/.julia/v0.6/Knet/examples/mnist.jl:117
 in (::MNIST.#kw##minibatch)(::Array{Any,1}, ::MNIST.#minibatch, ::Array{UInt8,1}, ::Array{UInt8,1}, ::Int64) at ./<missing>:0
 in main(::Array{String,1}) at /localhome/paulito/.julia/v0.6/Knet/examples/mnist.jl:43
 in main() at /localhome/paulito/.julia/v0.6/Knet/examples/mnist.jl:20

**Comment by denizyuret** [comment]
It does not look like it is able to locate the libknet8 library.  Is it in
the Pkg.dir("Knet/src") directory?  What is the value of the const libknet8?

deniz

**Comment by ppalmes** [comment]
yes, libknet8.so and libknet.so are in the Pkg.dir("Knet/src")

the value of Knet.libknet8 is ""

**Comment by ppalmes** [comment]
ok, i got it working now.

i have to compile (using make) manually in the Knet/src and Knet/src7. I think the Pkg.build("Knet") lacks the command to compile src7 directory.

**Comment by denizyuret** [comment]
src7 is deprecated, I am only keeping it around temporarily, no need to
compile there...

**Comment by ppalmes** [comment]
is there a manual way to make sure libknet8 will be available to Knet Julia? the libs are there so I think we just have to export the location?

**Comment by denizyuret** [comment]
Normally the info under Knet/deps should cause the compilation libknet8
automatically when you install and when you run Pkg.build("Knet").  Then
the package looks for it under Pkg.dir("Knet/src") -- see the first line of
"src/gpu.jl".  Maybe your Knet is not at a standard location?

**Comment by ppalmes** [comment]
i just followed the steps:

Pkg.checkout("Knet")
Pkg.build("Knet")
and the location of Knet is: /localhome/paulito/.julia/v0.6/Knet
which I think is correct.

**Comment by denizyuret** [comment]
Hmm, I should test with 0.6 as well.  Is there a downloadable version
somewhere or did you compile?

**Comment by ppalmes** [comment]
let me send the link. viral sent us the link but i'm afraid he sent us binary version for powerpc machines. let me check the email...

**Comment by ppalmes** [comment]
here's a  snapshot of the email from Viral. I think the first link is the binary version for ppc while the second link is the source code.

Can you try this one? I believe this error is addressed now.
http://s3.amazonaws.com/julianightlies/bin/linux/ppc64le/julia-latest-linuxppc64.tar.gz

Also, Elliot enabled the nightlies:
https://build.julialang.org/builders/package_tarballppc64le

**Comment by denizyuret** [comment]
One of my students had the same problem -- apparently the precompiled
Knet.ji file is the culprit.  If you remove that file it recompiles it with
the right path and it should work.

**Comment by ppalmes** [comment]
thanks. 

I removed Knet.ji  but got different error:

julia> using Knet
INFO: Precompiling module Knet.
julia> include(Pkg.dir("Knet/examples/mnist.jl"))
INFO: Loading MNIST...
MNIST
julia> gpu(true)
0
julia> MNIST.main()
opts=(:seed,-1)(:batchsize,100)(:hidden,Int64[])(:epochs,10)(:lr,0.5)(:atype,"KnetArray{Float32}")(:winit,0.1)(:fast,false)
(:epoch,0,:trn,0.06503333f0,:tst,0.0552f0)
ERROR: MethodError: no method matching tofloat(::Knet.KnetArray{Float32,2})
Closest candidates are:
  tofloat(::Number) at /localhome/paulito/.julia/v0.6/AutoGrad/src/core.jl:544
  tofloat{T<:Number}(::AbstractArray{T<:Number,N}) at /localhome/paulito/.julia/v0.6/AutoGrad/src/core.jl:545
  tofloat(::Tuple) at /localhome/paulito/.julia/v0.6/AutoGrad/src/core.jl:547
  ...
 in map(::AutoGrad.#tofloat, ::Array{Any,1}) at ./essentials.jl:124
 in tofloat(::Array{Any,1}) at /localhome/paulito/.julia/v0.6/AutoGrad/src/core.jl:549
 in forward_pass(::Function, ::Tuple{Array{Any,1},Knet.KnetArray{Float32,2},Knet.KnetArray{Float32,2}}, ::Array{Any,1}, ::Int64) at /localhome/paulito/.julia/v0.6/AutoGrad/src/core.jl:68
 in (::AutoGrad.##gradfun#1#3{MNIST.#loss,Int64})(::Array{Any,1}, ::Function, ::Array{Any,1}, ::Vararg{Any,N}) at /localhome/paulito/.julia/v0.6/AutoGrad/src/core.jl:37
 in (::AutoGrad.#gradfun#2)(::Array{Any,1}, ::Vararg{Any,N}) at /localhome/paulito/.julia/v0.6/AutoGrad/src/core.jl:37
 in #train#3(::Float64, ::Int64, ::Function, ::Array{Any,1}, ::Array{Any,1}) at /localhome/paulito/.julia/v0.6/Knet/examples/mnist.jl:79
 in (::MNIST.#kw##train)(::Array{Any,1}, ::MNIST.#train, ::Array{Any,1}, ::Array{Any,1}) at ./<missing>:0
 in macro expansion at /localhome/paulito/.julia/v0.6/Knet/examples/mnist.jl:51 [inlined]
 in macro expansion at ./util.jl:184 [inlined]
 in main(::Array{String,1}) at /localhome/paulito/.julia/v0.6/Knet/examples/mnist.jl:50
 in main() at /localhome/paulito/.julia/v0.6/Knet/examples/mnist.jl:20

**Comment by denizyuret** [comment]
Can you do a Pkg.update(), I don't think you have the latest AutoGrad
version.  (Should use v0.0.2).

**Comment by ppalmes** [comment]
MNIST.main() with gpu(true) works now with the latest AutoGrad.

I timed gpu(false) vs gpu(true) for MNIST.main().

They finished both around 3 to 4 seconds. It feels no difference between the two. Even the printing of results look similar speed.  

By the way, latest  Knet no longer uses compile and other macros. Is there a documentation for the latest Knet? 

Thanks.

**Comment by denizyuret** [comment]
I am writing the documentation as we speak!!!

In summary, Knet8 does not compile models any more, it uses AutoGrad to
differentiate ordinary Julia code!

When I test with MNIST.main("--fast") I get 2.82 for the cpu and 2.12 for
the gpu.  (--fast disables the loss printing and gets a better estimate of
real training cost).  This is 10 epochs on a softmax model.  To see some
more difference you can try MNIST.main("--fast --hidden 64") for a single
hidden layer with 64 units.  I get 6.23 with the cpu and 2.96 with the gpu.

best,
deniz

**Comment by ppalmes** [comment]
thanks for the explanation.

I'm just trying to make sure that GPU is really working ;).

it is great to have Knet GPU working now in Power. So far, this is the only deep learning package easy to install relative to other deep learning packages I tried and I very much prefer the programming logic of Knet. the documentation of the previous version is great and really looking forward to use Knet in my workflow. thanks for the great package!!!

**Comment by denizyuret** [comment]
You are very welcome.  You can sign up on the knet-users mailing list (very
low volume) to get updates as I try to finish v0.8.

BTW, I couldn't compile Julia 6 on the intel servers I have (Centos 6.5)
and the nightlies seem to be built for a different architecture.  Let me
know if you come across an image that might work.

deniz

**Comment by ppalmes** [comment]
I'll ask Viral if he has Intel binaries also. 

I like to contribute someday on the aspects of running Knet in PowerPC  machines. I like the simplicity of Knet which is true to the KISS principles of UNIX ;)

**Comment by ppalmes** [comment]
by the way, the latest version REQUIRE commented out CUDArt, ArgParse, and GZip which are needed to run MNIST.main()

**Comment by denizyuret** [comment]
Yes, I decided to minimize the requirements to run everything under src and
write your own models.  I will add something to the documentation of
additional requirements to run examples, or use the gpu.

**Comment by AStupidBear** [comment]
@denizyuret Thank you! Knet8 is usable for me. 
Will knet7 and knet8 be developed in parallel or knet8 will be a successor of knet7? If it's the latter case, can AutoGrad emulate the performance of knet7 on gpu?

**Comment by denizyuret** [comment]
knet7 will be maintained with bugfixes, future work will focus on knet8.
I am working on knet8 performance on the gpu, hope to make it better than
knet7.
